python/caffe/io.py

```diff
@@ -20,11 +20,20 @@ def blobproto_to_array(blob, return_diff=False):
     Convert a blob proto to an array. In default, we will just return the data,
     unless return_diff is True, in which case we will return the diff.
     """
+    # Read the data into an array
     if return_diff:
-        return np.array(blob.diff).reshape(*blob.shape.dim)
+        data = np.array(blob.diff)
     else:
-        return np.array(blob.data).reshape(*blob.shape.dim)
+        data = np.array(blob.data)
 
+    if blob.HasField('shape'):
+        return data.reshape(blob.shape.dim)
+    elif blob.HasField('num') or blob.HasField('channels') or blob.HasField('height') or blob.HasField('width'):
+        # Use legacy 4D shape
+        return data.reshape(blob.num, blob.channels, blob.height, blob.width)
+    else:
+        # Return scalar
+        return data
```

Comment:
This wasn't quite what I meant -- this will give a 1D array, not a 0D array (scalar).  You'd still want to `return data.reshape(blob.shape.dim)`, or equivalently `return data.reshape(())`.  The invariant we want is `len(data.shape) == len(blob.shape.dim)` (unless any of the legacy fields are present).  It will only work in the case where `len(data) == 1`, but that's okay since a `BlobProto` with no `shape` field (and no legacy dims) is kind of an odd corner case.